nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java

```diff
@@ -181,133 +185,223 @@ public IcebergTable table(TableIdentifier tableIdentifier) {
       throws NoSuchNamespaceException {
+    ContentKey key = ContentKey.of(namespace.levels());
     try {
-      return withReference(
-              getApi()
-                  .getNamespace()
-                  .namespace(org.projectnessie.model.Namespace.of(namespace.levels())))
-          .get()
-          .getProperties();
-    } catch (NessieNamespaceNotFoundException e) {
-      throw new NoSuchNamespaceException(e, "Namespace does not exist: %s", namespace);
-    } catch (NessieReferenceNotFoundException e) {
+      Map<ContentKey, Content> contentMap = withReference(api.getContent()).key(key).get();
+      Content existing = contentMap.get(key);
+      if (!(existing instanceof org.projectnessie.model.Namespace)) {
+        throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+      }
+      return ((org.projectnessie.model.Namespace) existing).getProperties();
+    } catch (NessieNotFoundException e) {
       throw new RuntimeException(
           String.format(
-              "Cannot load Namespace '%s': " + "ref '%s' is no longer valid.",
+              "Cannot load Namespace '%s': ref '%s' is no longer valid.",
               namespace, getRef().getName()),
           e);
     }
   }
 
   public boolean setProperties(Namespace namespace, Map<String, String> properties) {
-    try {
-      withReference(
-              getApi()
-                  .updateProperties()
-                  .namespace(org.projectnessie.model.Namespace.of(namespace.levels()))
-                  .updateProperties(properties))
-          .update();
-      refresh();
-      // always successful, otherwise an exception is thrown
-      return true;
-    } catch (NessieNamespaceNotFoundException e) {
-      throw new NoSuchNamespaceException(e, "Namespace does not exist: %s", namespace);
-    } catch (NessieNotFoundException e) {
-      throw new RuntimeException(
-          String.format(
-              "Cannot update properties on Namespace '%s': ref '%s' is no longer valid.",
-              namespace, getRef().getName()),
-          e);
-    }
+    return updateProperties(namespace, props -> props.putAll(properties));
   }
 
   public boolean removeProperties(Namespace namespace, Set<String> properties) {
+    return updateProperties(namespace, props -> props.keySet().removeAll(properties));
+  }
+
+  private boolean updateProperties(Namespace namespace, Consumer<Map<String, String>> action) {
+    getRef().checkMutable();
+    ContentKey key = ContentKey.of(namespace.levels());
+
     try {
-      withReference(
-              getApi()
-                  .updateProperties()
-                  .namespace(org.projectnessie.model.Namespace.of(namespace.levels()))
-                  .removeProperties(properties))
-          .update();
-      refresh();
+
+      Map<ContentKey, Content> contentMap =
+          api.getContent().reference(getReference()).key(key).get();
```

Comment:
From the new code, I guess the intention is to load the _latest_ properties, is it not? Is that case, I guess we may want to refresh the reference before loading... WDYT?

Reply:
The code here is _updating_ the properties, not _loading_ them. 

In `loadNamespaceMetadata` there is indeed a call to `withReference` that will omit the current hash if the ref is mutable, thus resulting in loading the latest data from the HEAD of that ref.

Here, however, I think it's important to update the content by specifying the latest hash we know about, so that any conflict that might happen due to more recent commits will surface as `NessieConflictException` – not `NessieBadRequestException`.

Reply:
Actually, I think we may have a deeper problem here. The `commitRetry` call (below) can advance (refresh) the reference without rebuilding the property set. I think this is incorrect. Every time the (local) HEAD advances we have to reload the namespace properties from the Nessie server and re-apply changes. WDYT?

Reply:
In other words the commit to update Namespace properties should set the "expected hash" to the value that was used for loading the properties before the modification.

Reply:
Cf. https://github.com/projectnessie/nessie/blob/81b34b610897c6f54bca88e8e6d503cfc3bbe05f/api/client/src/main/java/org/projectnessie/client/api/ns/ClientSideUpdateNamespace.java#L56

Reply:
> the commit to update Namespace properties should set the "expected hash" to the value that was used for loading the properties before the modification.

Isn't that being done already? `.reference(getReference())` is the same as `.refName(getReference().getName()).hashOnRef(getReference().getHash())`.

>  Every time the (local) HEAD advances we have to reload the namespace properties from the Nessie server and re-apply changes. 

I'm not sure it's necessary?  If the remote head of the branch has moved, either there is a conflict, in which case we don't retry and the method will throw; or there is no conflict, in which case the first cmmit attempt will succeed. These methods don't return the modified namespace btw, only a boolean; so I don't see why we need to reload the namespace content between each retry.

Reply:
This may not manifest in practice, but `commitRetry` has code that can advance the reference. If the reference advances and a re-try occurs, the server-side namespace properties may have entries that are not in the updated properties parameter... since the client does not know about the new server-side properties, they will be lost on re-try. However, Iceberg javadoc indicates that properties not mentioned in the update set should be retained. Note that the `commitRetry` will use the new HEAD without reloading Namespace properties, so it will make the impression on the server side that the total property set replacement is intentional.

I think this is actually the use case where we'd want to re-try on conflict. However, I believe the re-try logic should get the latest properties and re-apply the property updates (thus incorporating possible unrelated property changes). WDYT?

Reply:
> I think this is actually the use case where we'd want to re-try on conflict. However, I believe the re-try logic should get the latest properties and re-apply the property updates (thus incorporating possible unrelated property changes)

Agreed: _depending on what an update actually does to the content object_, it might make sense to retry in case of conflicts. IMHO for creates and deletes it never does. But here, for instance, we could consider retrying conflicts, provided that we refresh the content to update between each retry.

I will make the necessary changes to `commitRetry`.

Reply:
Done in 8c8fbfa.